src/Microsoft.AspNet.Proxy/ProxyMiddleware.cs

```diff
@@ -0,0 +1,89 @@
+        private readonly HttpClient _httpClient;
+        private ProxyOptions _options;
+
+        public ProxyMiddleware([NotNull] RequestDelegate next, [NotNull] ProxyOptions options)
+        {
+            _next = next;
+            if (string.IsNullOrEmpty(options.Host))
+            {
+                throw new InvalidOperationException("Options parameter must specify host");
+            }
+
+            // Setting default Port and Scheme if not specified
+            if (string.IsNullOrEmpty(options.Port))
+            {
+                options.Port = "80";
+            }
+
+            if (string.IsNullOrEmpty(options.Scheme))
+            {
+                options.Scheme = "http";
+            }
+
+            _options = options;
+            _httpClient = new HttpClient();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var requestMessage = new HttpRequestMessage();
+            if (string.Equals(context.Request.Method, "POST", StringComparison.OrdinalIgnoreCase))
+            {
+                var proxyStreamContent = new ProxyStreamContent(context.Request.Body);
+                requestMessage.Content = proxyStreamContent;
+                requestMessage.Method = new HttpMethod("POST");
+            }
+
+            // Copy the request headers
+            foreach (KeyValuePair<string, string[]> header in context.Request.Headers)
+            {
+                if (!requestMessage.Headers.TryAddWithoutValidation(header.Key, header.Value))
+                {
+                    bool success = requestMessage.Content.Headers.TryAddWithoutValidation(header.Key, header.Value);
+                }
+            }
+
+            requestMessage.Headers.Host = _options.Host + ":" + _options.Port;
+            var uriString = $"{_options.Scheme}://{_options.Host}:{_options.Port}{context.Request.PathBase}{context.Request.Path}{context.Request.QueryString}";
+
+            requestMessage.RequestUri = new Uri(uriString);
+            requestMessage.Method = new HttpMethod(context.Request.Method);
+
+            var responseMessage = await _httpClient.SendAsync(requestMessage);
+            context.Response.StatusCode = (int)responseMessage.StatusCode;
+            foreach (KeyValuePair<string, IEnumerable<string>> header in responseMessage.Headers)
+            {
+                context.Response.Headers.SetValues(header.Key, header.Value.ToArray());
+            }
+
+            foreach (KeyValuePair<string, IEnumerable<string>> header in responseMessage.Content.Headers)
+            {
+                context.Response.Headers.SetValues(header.Key, header.Value.ToArray());
+            }
+
+            // SendAsync doesn't chunck response. This removes the header that tells context to expected chunked data.
+            context.Response.Headers.Remove("transfer-encoding");
+            await responseMessage.Content.CopyToAsync(context.Response.Body);
+            context.Dispose();
```

Comment:
Everything after SendAsync should be in  try block and Dispose should be in a finally block, or else you could leak connections.